docs/pages/pmd/rules/java/errorprone.md

```diff
@@ -2850,16 +2854,26 @@ with the restriction that the logger needs to be passed into the constructor.
 
 **This rule is defined by the following XPath expression:**
 ``` xpath
-//ClassOrInterfaceBodyDeclaration[FieldDeclaration//ClassOrInterfaceType[@Image='Log']
- and
- not(FieldDeclaration[@Final='true'][@Static='true'][@Private='true'][.//VariableDeclaratorId[@Image=$staticLoggerName]]
- and
- //ArgumentList//ClassOrInterfaceType[@Image = ancestor::ClassOrInterfaceDeclaration/@Image or @Image = ancestor::EnumDeclaration/@Image])
- and
- not(FieldDeclaration[@Final='true'][@Private='true'][.//VariableDeclaratorId[@Image='log']]
- [count(.//VariableInitializer)=0]
- [ancestor::ClassOrInterfaceBody//StatementExpression[.//PrimaryExpression/descendant::*[@Image='log']][count(.//AllocationExpression)=0]]
- )]
+//ClassOrInterfaceBodyDeclaration[
+ FieldDeclaration//ClassOrInterfaceType[pmd-java:typeIs($loggerClass)]
+  and not (
+   (: check logger name :)
+   (
+    FieldDeclaration[$requireStaticFinalLogger][@Final=true()][@Static=true()][.//VariableDeclaratorId[@Image=$loggerName or @Image=$staticLoggerName]]
+    |
+    FieldDeclaration[$requireStaticFinalLogger = false()][.//VariableDeclaratorId[@Image=$loggerName or @Image=$staticLoggerName]]
+   )
+   and
+   (: in place initialization with method argument pointing to current class (even for Enums) :)
+   .//ArgumentList//ClassOrInterfaceType[@Image = ancestor::ClassOrInterfaceDeclaration/@Image or @Image = ancestor::EnumDeclaration/@Image]
+  )
+  and not (
+   (: final logger initialized inside constructor :)
+   FieldDeclaration[@Final=true()][@Private=true()][.//VariableDeclaratorId[@Image=$loggerName or @Image=$staticLoggerName]]
+   [count(.//VariableInitializer)=0]
+   [ancestor::ClassOrInterfaceBody//StatementExpression[.//PrimaryExpression/descendant::*[@Image=$loggerName or @Image=$staticLoggerName]][count(.//AllocationExpression)=0]]
```

Comment:
I have a question about this `[count(.//AllocationExpression)=0]` part, when I remove it current tests are all green. What is it used for? Why creation new instance of something is considered as bug (there is not a test case for it)?
And I found another use-case that could be checked
```java
public class Xxx {
    private final Log log;
    Xxx() {
        this.log = MyLoggerFactory.getLogger(Xxx.class);
    }
}
  
public class Xxx {
    private final Log log;
    Xxx() {
        this.log = MyLoggerFactory.getLogger(Bar.class);
    }
}
```

Reply:
I guess, the original author wanted to prevent things like this:

`private final Log log = new Log(MyClass.class)`

Usually, the logging framework provide factories, so that the logger instances are shared singletons and under control by the logging framework. With explicitly calling a constructor, we would not use the logging framework correctly (hopefully, the constructor is anyway not public, so this would be theoretical problem). And yes, if no test is failing, we have no test case for this.

Your other use case: Is the problem, you see here, that the logger for Bar is used in the class Xxx? Meaning, you first snippet is the correct one (no violation), while we should flag the second snippet? Or is the issue with initializing the field in the constructor? Actually, I would prefer to initialize the (final) field directly in the declaration....